### PR TITLE
CFC-121 Create ulx curse and ulx timedcurse

### PR DESCRIPTION
**Pull request by legokidlogan**

### CODE REVIEW

File: lua/cfc_ulx_commands/curse/effects/ball.lua [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    minDuration = nil,
+    maxDuration = nil,
+    onetimeDurationMult = nil,
+    excludeFromOnetime = true,
+} )
+
+
+if CLIENT then return end
+
+hook.Add( "CFC_ULXCommands_Balls_CanUnball", HOOK_PREFIX .. "BlockUnball", function( ply )
```
Comment (wrefgtzweve) [inline lua/cfc_ulx_commands/curse/effects/ball.lua:37]: Does this hook need to exist when the effect isn't being used?

File: lua/cfc_ulx_commands/curse/effects/disable_noclip.lua [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    end,
+
+    minDuration = nil,
+    maxDuration = nil,
+    onetimeDurationMult = nil,
+    excludeFromOnetime = true,
+} )
+
+
+hook.Add( "PlayerNoClip", HOOK_PREFIX .. "BlockNoclip", function( ply, desiredState )
```
Comment (wrefgtzweve) [inline lua/cfc_ulx_commands/curse/effects/disable_noclip.lua:34]: Hook should only exist when effect exists

File: lua/cfc_ulx_commands/curse/sv_utils.lua [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    ply.CFCUlxCurseEffect = nil
+    ply.CFCUlxCurseEffectExpireTime = nil
+    prevEffect.onEnd( ply )
+end )
+
+
+timer.Create( "CFC_ULXCommands_Curse_StartAndStopEffects", 5, 0, function()
+    local now = RealTime()
+
+    for _, ply in ipairs( player.GetAll() ) do
```
Comment (wrefgtzweve) [inline lua/cfc_ulx_commands/curse/sv_utils.lua:72]: Should use a lookup table of all players with effects, not all players on the server.

File: lua/ulib/modules/client/cfc_timed_punishments.lua [hunk 4]
```diff
@@ -26,7 +26,8 @@ local function alert()
 end
 
 net.Receive( "CFC_TimedPunishments_Punishments", function()
-    punishments = {}
+    table.Empty( punishments )
```
Comment (wrefgtzweve) [inline lua/ulib/modules/client/cfc_timed_punishments.lua:29]: Is this related to the effects at all?

File: lua/cfc_ulx_commands/curse/effects/aim_sensitivity.lua [hunk 5]
```diff
@@ -0,0 +1,39 @@
+local EFFECT_NAME = "AimSensitivity"
```
Comment (wrefgtzweve) [inline lua/cfc_ulx_commands/curse/effects/aim_sensitivity.lua:1]: Should use hook `AdjustMouseSensitivity` for this probably

File: lua/cfc_ulx_commands/curse/effects/crouch.lua [hunk 6]
```diff
@@ -0,0 +1,26 @@
+local EFFECT_NAME = "Crouch"
```
Comment (wrefgtzweve) [inline lua/cfc_ulx_commands/curse/effects/crouch.lua:1]: Might be a bit boring, something like `cmd:CommandNumber() % 4 == 0` would probably spice it up

File: lua/cfc_ulx_commands/curse/effects/respawn.lua [hunk 7]
```diff
@@ -0,0 +1,20 @@
+local EFFECT_NAME = "Respawn"
+
+CFCUlxCurse.RegisterEffect( {
+    name = EFFECT_NAME,
+
+    onStart = function( cursedPly )
+        if CLIENT then return end
+
+        cursedPly:Spawn()
```
Comment (wrefgtzweve) [inline lua/cfc_ulx_commands/curse/effects/respawn.lua:9]: [KillSilent](https://wiki.facepunch.com/gmod/Player:KillSilent) might me more evil as they'll actually die, could run a timer that randomly triggers it too

File: lua/cfc_ulx_commands/curse/effects/weapon_indecision.lua [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+            if wepCount == 0 then return end
+
+            local wep = weps[math.random( wepCount )]
+
+            cursedPly:SelectWeapon( wep:GetClass() )
+        end )
+    end,
+
+    onEnd = function( cursedPly )
+        timer.Remove( HOOK_PREFIX .. "SelectWeapon_" .. cursedPly:SteamID64() )
```
Comment (wrefgtzweve) [inline lua/cfc_ulx_commands/curse/effects/weapon_indecision.lua:30]: Timers like these should be properly cleaned up, what if cursedPly is no longer valid at this point?

image

File: lua/cfc_ulx_commands/curse/sv_utils.lua [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+    local prevEffect = CFCUlxCurse.GetCurrentEffect( ply )
+    if not prevEffect then return end
+
+    ply.CFCUlxCurseEffect = nil
+    prevEffect.onEnd( ply )
+end )
+
+
+timer.Create( "CFC_ULXCommands_Curse_StartAndStopEffects", 5, 0, function()
+    local now = RealTime()
```
Comment (wrefgtzweve) [inline lua/cfc_ulx_commands/curse/sv_utils.lua:95]: You should use CurTime here as curtime properly adjusts for lag

File: lua/cfc_ulx_commands/curse/effects/respawn.lua (hunk 7 above)
Comment (legokidlogan) [reply lua/cfc_ulx_commands/curse/effects/respawn.lua:9]: I feel like it's more cursed to just randomly get teleported to spawn with no indication of why, especially since `:KillSilent()` causes a specific death 'animation' to play out that isn't seen anywhere else and thus makes it very identifiable.

You did give me an idea for another type of effect to add, though...

**Review by brandonsturgeon — APPROVED** [review]
